### PR TITLE
Gaussian Mixture Models

### PR DESCRIPTION
**Pull request by t0maboro**

### CODE REVIEW

File: src/algorithms/clustering/gmm.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        self.num_clusters = num_clusters
+        self.rows = self.df.shape[0]
+        self.dim = len(self.df.columns)
+        self.labels = np.floor_divide(np.arange(self.rows), self.rows / self.num_clusters)
+        self.mu_arr = self.initialize_mu()
+        self.sigma_arr = self.initialize_sigma()
+        self.pi_arr = self.initialize_pi()
+        self.prob_matrix = None
+        self.eps = float(eps)
+        self.max_iter = max_iterations or 99999999
```
Comment (mhawryluk) [inline src/algorithms/clustering/gmm.py]: is it intentional to not have an infinite loop, but rather a very big limit or was it just the easiest way to implement an "infinite" loop? it seems a little odd, especially this number looks very suspicious haha. please at least introduce a constant like `DEFAULT_MAX_ITER` to make it look more intentional (and maybe change the number to a more intentional too)

File: src/algorithms/clustering/gmm.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import numpy as np
+from PyQt5.QtWidgets import QMessageBox
+from numpy.linalg import LinAlgError
+from scipy.stats import multivariate_normal
```
Comment (gregori0o) [inline src/algorithms/clustering/gmm.py:4]: Please add scipy to requirements.txt file

File: src/algorithms/clustering/gmm.py (hunk 1 above)
Comment (gregori0o) [inline src/algorithms/clustering/gmm.py]: 'no limit' settings has a limit. Isn't it better to have real no limit?

File: src/widgets/options_widgets/gmm_options.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        self.num_clusters_spinbox.setValue(3)
+        self.layout.addRow(QLabel("Number of clusters:"), self.num_clusters_spinbox)
+
+        self.precision_box = QComboBox()
+        self.precision_box.addItem('1e-10')
+        self.precision_box.addItem('1e-8')
+        self.precision_box.addItem('1e-6')
+        self.precision_box.addItem('1e-4')
+        self.precision_box.addItem('1e-2')
+        self.precision_box.addItem('1')
```
Comment (gregori0o) [inline src/widgets/options_widgets/gmm_options.py]:
```suggestion
        self.precision_box.addItems(['1e-10', '1e-8', '1e-6', '1e-4', '1e-2', '1'])
```

File: src/visualization/clustering/gmm_vis.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+            self.control_buttons_layout.addWidget(self.right_box)
+            self.control_buttons_layout.addWidget(self.right_button)
+
+            self.visualization_box_layout.addLayout(self.control_buttons_layout, 0)
+        else:
+            self.step_label = QLabel("STEP: {}".format(self.current_step))
+            self.visualization_box_layout.addWidget(self.step_label, 0, alignment=Qt.AlignCenter)
+
+        # description
+        description = "Gaussian Mixture Methods algorithm - steps visualization."
```
Comment (gregori0o) [inline src/visualization/clustering/gmm_vis.py]: I think, the description should be longer and describe what we see in the visualization.

Comment (gregori0o) [reply src/visualization/clustering/gmm_vis.py]: Maybe we can improve descriptions of all algorithms in other PR

**Review by gregori0o — COMMENTED** [review]
Great job!

File: src/algorithms/clustering/gmm.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    def update_mu(self):
+        sums = [np.sum((self.df.T * self.prob_matrix[:, c]).T, axis=0) for c in range(self.num_clusters)]
+        sums = [sums[c] / (self.pi_arr[c] * self.rows) for c in range(self.num_clusters)]
+        self.mu_arr = np.array(sums)
+
+    def update_sigma(self):
+        for c in range(self.num_clusters):
+            new_sigma_component = np.zeros((self.dim, self.dim))
+            for i in range(self.rows):
+                data_row = np.array(self.df.iloc[i])-self.mu_arr[c]
```
Comment (mhawryluk) [inline src/algorithms/clustering/gmm.py]:
```suggestion
                data_row = np.array(self.df.iloc[i]) - self.mu_arr[c]
```

File: src/visualization/clustering/gmm_vis.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+import numpy as np
+from PyQt5.QtCore import Qt
+from PyQt5.QtWidgets import QWidget, QHBoxLayout, QVBoxLayout, QGroupBox, QFormLayout, QLabel, QSpinBox, QPushButton, \
+    QComboBox, QScrollArea, QSizePolicy
+from algorithms import get_samples
+from matplotlib import pyplot as plt, transforms
+from matplotlib.animation import FuncAnimation
+from matplotlib.backends.backend_qtagg import FigureCanvasQTAgg
+
```
Comment (mhawryluk) [inline src/visualization/clustering/gmm_vis.py]:
```suggestion
```

File: src/visualization/clustering/gmm_vis.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        self.axes.set_ylabel(name_y)
+        self.axes.set_xlim(min_x, max_x)
+        self.axes.set_ylim(min_y, max_y)
+        self.axes.scatter(x=vector_x, y=vector_y)
+        if drawing:
+            self.draw()
+        if self.animation:
+            return self.axes.collections
+
+    def _draw_variance(self, mean, sigma, label, max_label, n_std=2.0):
```
Comment (mhawryluk) [inline src/visualization/clustering/gmm_vis.py:32]: I don't see the ellipses when running animation :/ It is visible for step-by-step version, but not animation for some reason. this method is being called just fine, so I'm not sure what is the reason. Maybe the ellipses don't fit on my canvas? is it visible on your computer?

**Review by mhawryluk — COMMENTED** [review]
Great job! there is a lot of duplicated code from KMeans, but we will probably be doing a refactor as part of the "make new algorithms easy to add" user story

File: src/visualization/clustering/gmm_vis.py (hunk 7 above)
Comment (t0maboro) [reply src/visualization/clustering/gmm_vis.py:32]: Oh, I missed that case, will be fixed asap

**Review by gregori0o — APPROVED** [review]
🚀🚀🚀

**Review by mhawryluk — APPROVED** [review]
Well done! 🎉

**Pull request merged**